### PR TITLE
used regular expression in `format_is_iso()`

### LINKED ISSUE #50465: PERF use regular expression in format_is_iso?
Suggested here: https://github.com/pandas-dev/pandas/pull/50242#discussion_r1057825783

Here are the relevant lines of code:

https://github.com/pandas-dev/pandas/blob/4e88c3f0389a1a7ecff8bfd23615c84ed92db951/pandas/_libs/tslibs/parsing.pyx#L821-L838

### PR DESCRIPTION
**Pull request by ShashwatAgrawal20**
- [x] closes #50465 (Replace xxxx with the GitHub issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/v2.0.0.rst` file if fixing a bug or adding a new feature.

Added regular expressions in the `format_is_iso()` function 

``` python 
def format_is_iso(f: str) -> bint:
    """
    Does format match the iso8601 set that can be handled by the C parser?
    Generally of form YYYY-MM-DDTHH:MM:SS - date separator can be different
    but must be consistent.  Leading 0s in dates and times are optional.
    """
    iso_regex = \
        r"^\d{4}(-\d{2}(-\d{2}(T\d{2}:\d{2}:\d{2}(\.\d+)?(([+-]\d{2}:\d{2})|Z)?)?)?)?$"
    excluded_formats = ["%Y%m%d", "%Y%m", "%Y"]
    if re.match(iso_regex, f) and f not in excluded_formats:
        return True
    return False


```

### CODE REVIEW

File: pandas/_libs/tslibs/parsing.pyx [hunk 1]
```diff
@@ -824,17 +824,11 @@ def format_is_iso(f: str) -> bint:
     Generally of form YYYY-MM-DDTHH:MM:SS - date separator can be different
     but must be consistent.  Leading 0s in dates and times are optional.
     """
-    iso_template = "%Y{date_sep}%m{date_sep}%d{time_sep}%H:%M:%S{micro_or_tz}".format
+    iso_regex = \
+        r"^\d{4}(-\d{2}(-\d{2}(T\d{2}:\d{2}:\d{2}(\.\d+)?(([+-]\d{2}:\d{2})|Z)?)?)?)?$"
```
Comment (MarcoGorelli) [inline pandas/_libs/tslibs/parsing.pyx:828]: thanks for working on this - to keep it readable, how about making this a verbose  with comments?

see https://stackoverflow.com/a/13852096/4451315

Comment (ShashwatAgrawal20) [reply pandas/_libs/tslibs/parsing.pyx:828]: @MarcoGorelli added some comments for explaining the **regular expression** and how it's evaluating

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]
Please check the link I provided, it explains how to write a regular expression with the verbose flag

Did you try running this locally?

**Comment by ShashwatAgrawal20** [comment]
Oh sorry didn't read the lin,  let me fix it

File: pandas/_libs/tslibs/parsing.pyx [hunk 2]
```diff
@@ -824,17 +824,22 @@ def format_is_iso(f: str) -> bint:
     Generally of form YYYY-MM-DDTHH:MM:SS - date separator can be different
     but must be consistent.  Leading 0s in dates and times are optional.
     """
-    iso_template = "%Y{date_sep}%m{date_sep}%d{time_sep}%H:%M:%S{micro_or_tz}".format
+    iso_regex = re.compile(
+        r"""
+        ^                       # start of string
+        \d{4}                   # match a 4-digit year
+        (-\d{2})?               # optionally match a 2-digit month
```
Comment (MarcoGorelli) [inline pandas/_libs/tslibs/parsing.pyx:831]: `date_sep` isn't limited to `-`

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]

**Comment by ShashwatAgrawal20** [comment]
``` python 
import re

def format_is_iso(f: str) -> bool:
    """
    Does format match the iso8601 set that can be handled by the C parser?
    Generally of form YYYY-MM-DDTHH:MM:SS - date separator can be different
    but must be consistent.  Leading 0s in dates and times are optional.
    """
    iso_regex = re.compile(
        r"""
        ^                         # start of string
        \d{4}                     # match a 4-digit year
        (-\d{2})?                 # optionally match a 2-digit month
        (-\d{2})?                 # optionally match a 2-digit day
        (T\d{2}:\d{2}:\d{2}       # match time in the format "THH:MM:SS"
        (\.\d+)?                  # optionally match a decimal and fractional seconds
        ([+-]\d{2}:\d{2}|Z)?)?    # optionally match a timezone in the format "+HH:MM" or "Z"
        $                         # end of string
        """,
        re.VERBOSE,
    )
    excluded_formats = ["%Y%m%d", "%Y%m", "%Y"]
    if re.match(iso_regex, f) and f not in excluded_formats:
        return True
    return False


res = format_is_iso('2022-12-38T17:26:49Z')
print(res)

```

**Comment by ShashwatAgrawal20** [comment]
``` python 
        ([ -/\\.]\d{2}|\d{2})?  # optionally match a 2-digit month
        ([ -/\\.]\d{2}|\d{2})?  # optionally match a 2-digit day
        ([ T]\d{2}:\d{2}:\d{2}  # match time in the format "THH:MM:SS"

```

Added some separators regex to separate the **dates** & **time**

**Comment by MarcoGorelli** [comment]
getting there, but it's not quite right,it's failing lots of tests

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]
tests are still failing

at a minimum, you need to get
```
pytest pandas/tests/tslibs/test_parsing.py
```
to pass - please make sure that this passes locally for you

**Comment by ShashwatAgrawal20** [comment]
> tests are still failing
> 
> at a minimum, you need to get
> 
> ```
> pytest pandas/tests/tslibs/test_parsing.py
> ```
> 
> to pass - please make sure that this passes locally for you

Sorry about the inconvenience I didn't knew we can test this on local system

**Comment by ShashwatAgrawal20** [comment]
@MarcoGorelli ran the test on the **non-regular-expression** version they are also failing. What should I do? any suggestions? 
image

**Comment by ShashwatAgrawal20** [comment]
@MarcoGorelli ran the test on the **non-regular-expression** version they are also failing. What should I do? any suggestions? 
image

**Comment by MarcoGorelli** [comment]
did you compile the C extensions? https://pandas.pydata.org/docs/dev/development/contributing_environment.html#step-3-build-and-install-pandas

**Comment by ShashwatAgrawal20** [comment]
> @MarcoGorelli ran the test on the **non-regular-expression** version they are also failing. What should I do? any suggestions? image

Hey, @MarcoGorelli didn't understand why it's `assert False == True` in all. It should compare the expected value with the result value. 
As in the file `pandas/tests/tslibs/test_parsing.py`
```python 
@pytest.mark.parametrize(
    "fmt,expected",
    [
        ("%Y %m %d %H:%M:%S", True),
        ("%Y/%m/%d %H:%M:%S", True),
        (r"%Y\%m\%d %H:%M:%S", True),
        ("%Y-%m-%d %H:%M:%S", True),
        ("%Y.%m.%d %H:%M:%S", True),
        ("%Y%m%d %H:%M:%S", True),
        ("%Y-%m-%dT%H:%M:%S", True),
        ("%Y-%m-%dT%H:%M:%S%z", True),
        ("%Y-%m-%dT%H:%M:%S%Z", False),
        ("%Y-%m-%dT%H:%M:%S.%f", True),
        ("%Y-%m-%dT%H:%M:%S.%f%z", True),
        ("%Y-%m-%dT%H:%M:%S.%f%Z", False),
        ("%Y%m%d", False),
        ("%Y%m", False),
        ("%Y", False),
        ("%Y-%m-%d", True),
        ("%Y-%m", True),
    ],
)
def test_is_iso_format(fmt, expected):
    # see gh-41047
    result = parsing.format_is_iso(fmt)
    assert result == expected
```
Wrote the code with the dates in the format of the above and checked the result with the excepted output, the result output is matching with the expected out 

```python
import re


def format_is_iso(f: str) -> bool:
    """
    Does format match the iso8601 set that can be handled by the C parser?
    Generally of form YYYY-MM-DDTHH:MM:SS - date separator can be different
    but must be consistent.  Leading 0s in dates and times are optional.
    """
    iso_regex = re.compile(
        r"""
        ^                       # start of string
        \d{4}                   # match a 4-digit year
        ([ -/\\.]\d{2}|\d{2})?  # optionally match a 2-digit month
        ([ -/\\.]\d{2}|\d{2})?  # optionally match a 2-digit day
        ([ T]\d{2}:\d{2}:\d{2}  # match time in the format "THH:MM:SS"
        (\.\d+)?                # match a decimal and fractional seconds
        ([+-]\d{2}:\d{2}|Z)?)?  # match timezone in the format "+HH:MM" or "Z"
        $                       # end of string
        """,
        re.VERBOSE,
    )

    excluded_formats = [
        r"^\d{4}\d{2}\d{2}$",  # %Y%m%d
        r"^\d{4}\d{2}$",       # %Y%m
        r"^\d{4}$",            # %Y
    ]

    if any(re.match(pattern, f) for pattern in excluded_formats):
        return False

    return bool(re.match(iso_regex, f))


dates = [
    "2022 12 29 12:34:56",
    "2022/12/29 12:34:56",
    "2022\\12\\29 12:34:56",
    "2022-12-29 12:34:56",
    "2022.12.29 12:34:56",
    "20221229 12:34:56",
    "2022-12-29T12:34:56",
    "2022-12-29T12:34:56+01:00",
    "2022-12-29T12:34:56CET",
    "2022-12-29T12:34:56.123456",
    "2022-12-29T12:34:56.123456+01:00",
    "2022-12-29T12:34:56.123456CET",
    "20221229",
    "202212",
    "2022",
    "2022-12-29",
    "2022-12"
]

for date in dates:
    res = format_is_iso(date)
    print(date, res)

```
image

But at the time of `pytest` it fails again

**Comment by MarcoGorelli** [comment]
does the test pass on the `main` branch? if not, then I guess that your pandas build isn't quite correct - check the installation instructions I sent earlier

**Comment by ShashwatAgrawal20** [comment]
> does the test pass on the `main` branch? if not, then I guess that your pandas build isn't quite correct - check the installation instructions I sent earlier

The test got passed on the main branch with the `non-regular-expression`.

**Comment by MarcoGorelli** [comment]
then you need to fix up your regular expression 😄

**Comment by MarcoGorelli** [comment]
this might help https://stackoverflow.com/a/74953592/4451315

**Comment by ShashwatAgrawal20** [comment]
Hey the upstream got the function remove, don't know what exactly happend

**Comment by MarcoGorelli** [comment]
it moved to a different file, if you do a search (e.g. with `git grep`) you'll find it

**Comment by ShashwatAgrawal20** [comment]
> it moved to a different file, if you do a search (e.g. with `git grep`) you'll find it

this one ?

**Comment by MarcoGorelli** [comment]
yes

**Comment by ShashwatAgrawal20** [comment]
`pytest pandas/tests/tslibs/test_parsing.py`
do I still have to run this to test it locally? or has the test changed?

**Comment by MarcoGorelli** [comment]
yes, the tests need to pass

**Comment by ShashwatAgrawal20** [comment]
> yes, the tests need to pass

same test or did the **test** file got changed?

**Comment by MarcoGorelli** [comment]
it would take you less time to check than to ask 😄 should be the same file anyway

**Comment by ShashwatAgrawal20** [comment]
Hey, tried this to `using the main branch format_is_iso` code, to test what the output should come, as tried on some dates formats some and valid, and some are invalid, but it's showing false in every condition, wonder how it's passing the tests if it's showing false for true input, correct me if I am wrong 

``` python 
def format_is_iso(f: str) -> bool:
     excluded_formats = ["%Y%m"]
     for date_sep in [" ", "/", "\\", "-", ".", ""]:
         for time_sep in [" ", "T"]:
             for micro_or_tz in ["", "%z", ".%f", ".%f%z"]:
                iso_fmt = f"%Y{date_sep}%m{date_sep}%d{time_sep}%H:%M:%S{micro_or_tz}"
                if iso_fmt.startswith(f) and f not in excluded_formats:
                     return True
     return False


dates = [
    "2022 12 29 12:34:56",
    "2022/12/29 12:34:56",
    "2022\\12\\29 12:34:56",
    "2022-12-29 12:34:56",
    "2022.12.29 12:34:56",
    "20221229 12:34:56",
    "2022-12-29T12:34:56",
    "2022-12-29T12:34:56+01:00",
    "2022-12-29T12:34:56CET",
    "2022-12-29T12:34:56.123456",
    "2022-12-29T12:34:56.123456+01:00",
    "2022-12-29T12:34:56.123456CET",
    "20221229",
    "202212",
    "2022",
    "2022-12-29",
    "2022-12"
]

for date in dates:
    res = format_is_iso(date)
    print(date, res)
```

**Comment by MarcoGorelli** [comment]
the inputs to the test should be formats

https://github.com/pandas-dev/pandas/blob/502919e4b18e5d5aca584d4ebb84d663fcefffc9/pandas/tests/tslibs/test_parsing.py#L296-L315

**Comment by ShashwatAgrawal20** [comment]
The test got passed on the local system

File: pandas/_libs/tslibs/strptime.pyx [hunk 3]
```diff
@@ -60,15 +64,33 @@ cdef bint format_is_iso(f: str):
...
-            for micro_or_tz in ["", "%z", ".%f", ".%f%z"]:
-                iso_fmt = f"%Y{date_sep}%m{date_sep}%d{time_sep}%H:%M:%S{micro_or_tz}"
-                if iso_fmt.startswith(f) and f not in excluded_formats:
-                    return True
-    return False
+    iso_regex = re.compile(
+        r"""
+        ^                      # start of string
+        (?:                    # group for alternative date formats
+        %Y[ -/\\.]%m[ -/\\.]%d # 4-digit year, 2-digit month, 2-digit day
```
Comment (MarcoGorelli) [inline pandas/_libs/tslibs/strptime.pyx:71]: `date_sep` needs to be the same, e.g. `%Y-%m/%d` isn't allowed, but `%Y-%m-%d` and `%Y/%m/%d` is

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]
getting there 💪

File: pandas/_libs/tslibs/strptime.pyx (hunk 3 above)
Comment (ShashwatAgrawal20) [reply pandas/_libs/tslibs/strptime.pyx:71]: sorry, didn't get, what you mean but, 

- **%Y[ -/\\.]%m[ -/\\.]%d**: 4-d-year 2d-month & 2-d-day separated by (space, a hyphen, forward slash, or backward slash)
- **%Y%m%d**: same as above without the separators 
- **%Y[ -/\\.]%m**: 4-d-year & 2-d-month with separated by (space, a hyphen, forward slash, or backward slash)
- **%Y%m**: same without separators
- **%Y**: 4-digit year.

Comment (MarcoGorelli) [reply pandas/_libs/tslibs/strptime.pyx:71]: check the non-regex code, `date_sep` needs to be the same between %Y and %m and between %m and %d

`format_is_iso('%Y-%m/%d')` should return `False`

File: pandas/_libs/tslibs/strptime.pyx [hunk 4]
```diff
@@ -60,15 +64,37 @@ cdef bint format_is_iso(f: str):
...
+    iso_regex = re.compile(
+        r"""
+        ^                      # start of string
+        (?:                    # group for alternative date formats
+        %Y[-]%m[-]%d           # year month & date separated by either hyphen
+        | %Y[/]%m[/]%d         # OR separated by forward slash
+        | %Y[.]%m[.]%d         # OR separated by period
+        | %Y[ ]%m[ ]%d         # OR separated by space
+        | %Y[\\]%m[\\]%d       # OR separated by backslash
+        | %Y%m%d               # OR no separators
```
Comment (MarcoGorelli) [inline pandas/_libs/tslibs/strptime.pyx:76]: this is getting too long - can you use a backreference to match the first captured separator? https://www.pythontutorial.net/python-regex/python-regex-backreferences/

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]

File: pandas/_libs/tslibs/strptime.pyx (hunk 4 above)
Comment (ShashwatAgrawal20) [reply pandas/_libs/tslibs/strptime.pyx:76]: @MarcoGorelli what about this 
```python 
%Y([-/.\\ ])%m\1%d  # year month & date separated by either ([-/.\\ ])
```

Comment (MarcoGorelli) [reply pandas/_libs/tslibs/strptime.pyx:76]: maybe, try and see

Comment (ShashwatAgrawal20) [reply pandas/_libs/tslibs/strptime.pyx:76]: That worked

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]
This erroneously matches "%Y-%m %H:%M:%S", though it shouldn't (the original non-regex code, in fact, doesn't)

**Comment by ShashwatAgrawal20** [comment]
> This erroneously matches "%Y-%m %H:%M:%S", though it shouldn't (the original non-regex code, in fact, doesn't)

what about updating the exclude list like 
```python 
excluded_formats = [
        r"^%Y%m$",
        r"^%Y[/.\\ -]%m[ |T]%H:%M.*$",
        r"^%Y%m[ |T]%H:%M.*$",
]
```

**Comment by MarcoGorelli** [comment]
I was thinking more like
```
%Y
(?:[./\\-]?
  (?:%m
    (?:\1
      ?(:%d



      )
    )
  )
)
```

**Comment by ShashwatAgrawal20** [comment]
> I was thinking more something like
> 
> ```
> %Y
> (?:[./\\-]?
>   (?:%m
>     (?:\1
>       ?(:%d
> 
> 
> 
>       )
>     )
>   )
> )
> ```
> 
> with HMS in the middle

didn't think like that can try to work.

**Comment by ShashwatAgrawal20** [comment]
> I was thinking more something like
> 
> ```
> %Y
> (?:[./\\-]?
>   (?:%m
>     (?:\1
>       ?(:%d
> 
> 
> 
>       )
>     )
>   )
> )
> ```
> 
> with HMS in the middle

Like this ?

```python 
    iso_regex = re.compile(
        r"""
        ^                      # start of string
        %Y
        (?:([-/ \\.]?)
          (?:%m
            (?: \1
              (?: %d
                (?: [ |T]
                  (?: %H
                    (?: \:
                      (?: %M
                        (?: \:
                          (?: %S
                            (?: %z|.%f(?:%z|Z)?
                            )?
                          )?
                        )?
                      )?
                    )?
                  )?
                )?
              )?
            )?
          )?
        )?
        $                      # end of string
        """,
        re.VERBOSE,
    )

```

**Comment by MarcoGorelli** [comment]
Nice! you can probably simplify
```                    (?: \:
                      (?: %M
```
to something like
```
                    (?:\:%M
```
, given that the separator between %H and %M can only be `:`

**Comment by ShashwatAgrawal20** [comment]
> Nice! you can probably simplify
> 
> ```
>                       (?: %M
> ```
> 
> to something like
> 
> ```
>                     (?:\:%M
> ```
> 
> , given that the separator between %H and %M can only be `:`

What about this? 

```python 
    iso_regex = re.compile(
        r"""
        ^                     # start of string
        %Y                    # Year
        (?:([-/ \\.]?)%m      # month with or without separators
        (?:\1%d               # day with or without separators
        (?:[ |T]%H            # hour with separator
        (?:\:%M               # minute with separator
        (?:\:%S               # second with separator
        (?:%z|.%f(?:%z|Z)?    # timezone or fractional second
        )?)?)?)?)?)?          # optional
        $                     # end of string
        """,
        re.VERBOSE,
    )

```

**Comment by MarcoGorelli** [comment]
Nice! If that works, it looks good to me!

File: pandas/_libs/tslibs/strptime.pyx [hunk 5]
```diff
@@ -60,15 +64,27 @@ cdef bint format_is_iso(f: str):
...
-                iso_fmt = f"%Y{date_sep}%m{date_sep}%d{time_sep}%H:%M:%S{micro_or_tz}"
-                if iso_fmt.startswith(f) and f not in excluded_formats:
-                    return True
-    return False
+    iso_regex = re.compile(
+        r"""
+        ^                     # start of string
+        %Y                    # Year
+        (?:([-/ \\.]?)%m      # month with or without separators
+        (?:\1%d               # day with or without separators
```
Comment (MarcoGorelli) [inline pandas/_libs/tslibs/strptime.pyx:72]: perhaps let's update the comment to `"day with same separator as for year-month"`?

File: pandas/_libs/tslibs/strptime.pyx [hunk 6]
```diff
@@ -60,15 +64,27 @@ cdef bint format_is_iso(f: str):
...
+        $                     # end of string
+        """,
+        re.VERBOSE,
+    )
+    excluded_formats = [
+        r"^%Y%m$",
+    ]
+    if any(re.match(pattern, f) for pattern in excluded_formats):
+        return False
+    return bool(re.match(iso_regex, f))
```
Comment (MarcoGorelli) [inline pandas/_libs/tslibs/strptime.pyx:87]: I don't think we need a regex for the excluded formats - would
```python
return re.match(iso_regex, f) is not None and f not in excluded_formats
```
work?

File: pandas/_libs/tslibs/strptime.pyx [hunk 7]
```diff
@@ -43,13 +44,16 @@ from pandas._libs.tslibs.np_datetime cimport (
     pydatetime_to_dt64,
     string_to_dts,
 )
+
 from pandas._libs.tslibs.np_datetime import OutOfBoundsDatetime
+
```
Comment (MarcoGorelli) [inline pandas/_libs/tslibs/strptime.pyx:49]: could you revert these unrelated changes please?

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]
If this regex works, it looks good to me, thanks for sticking with it

Would be good to time it too

**Comment by ShashwatAgrawal20** [comment]
> Nice! If that works, it looks good to me!

hey, by the way, `pytest pandas/tests/tslibs/test_parsing.py` having some import error, is that on my device or it is actually an issue?

**Comment by MarcoGorelli** [comment]
it works for me, so my only guess is that it has to do with C extensions again - if you post your error (here or on slack) I'll take a look

File: pandas/_libs/tslibs/strptime.pyx (hunk 6 above)
Comment (ShashwatAgrawal20) [reply pandas/_libs/tslibs/strptime.pyx:87]: yea worked fine without the regex excluded_format

**Comment by ShashwatAgrawal20** [comment]
> it works for me, so my only guess is that it has to do with C extensions again - if you post your error (here or on slack) I'll take a look

Sorry for the inconvenience still a noob in the open source world, but learned a lot by your guidance 

image

**Comment by MarcoGorelli** [comment]
no worries - I suspect it's this https://github.com/pandas-dev/pandas/issues/47305
I'd suggest cleaning and rebuilding:
```
git clean -fxfd pandas
python setup.py build_ext -j 4
```

File: pandas/_libs/tslibs/strptime.pyx [hunk 8]
```diff
@@ -60,15 +62,23 @@ cdef bint format_is_iso(f: str):
     Generally of form YYYY-MM-DDTHH:MM:SS - date separator can be different
     but must be consistent.  Leading 0s in dates and times are optional.
     """
+    iso_regex = re.compile(
+        r"""
+        ^                     # start of string
+        %Y                    # Year
+        (?:([-/ \\.]?)%m      # month with or without separators
+        (?: \1%d              # day with same separator as for year-month
+        (?:[ |T]%H            # hour with separator
```
Comment (MarcoGorelli) [inline pandas/_libs/tslibs/strptime.pyx:71]: square brackets match anything inside them, you don't need `|`
```suggestion
        (?:[ T]%H             # hour with separator
```

File: pandas/_libs/tslibs/strptime.pyx [hunk 9]
```diff
@@ -60,15 +62,23 @@ cdef bint format_is_iso(f: str):
...
+        (?:\:%M               # minute with separator
+        (?:\:%S               # second with separator
+        (?:%z|.%f(?:%z|Z)?    # timezone or fractional second
```
Comment (MarcoGorelli) [inline pandas/_libs/tslibs/strptime.pyx:74]: the full stop needs escaping
```suggestion
        (?:%z|\.%f(?:%z|Z)?   # timezone or fractional second
```

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]
Cool, seems to be working!

Couple of minor comments - then I'll time this and we'll see if we can get it in

**Comment by MarcoGorelli** [comment]
Timings look:

upstream/main:
```python
In [2]: %%timeit
   ...: format_is_iso('%Y-%m-%d %H:%M%S%z')
   ...: 
   ...: 
2.64 µs ± 27.9 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
```

here:
```python
In [1]:  from pandas._libs.tslibs.strptime import format_is_iso

In [2]: %%timeit
   ...: format_is_iso('%Y-%m-%d %H:%M%S%z')
   ...: 
   ...: 
1.39 µs ± 67.7 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
```

(note: I had to change `cdef` to `cpdef` in both cases to be able to run this test)

I prefer it written like this anyway

**Review by MarcoGorelli — APPROVED** [review]
Looks good to me - @WillAyd seeing as you'd suggested this, any thoughts?

File: pandas/_libs/tslibs/strptime.pyx (hunk 8 above)
Comment (ShashwatAgrawal20) [reply pandas/_libs/tslibs/strptime.pyx]: should I directly commit the suggestions?

File: pandas/_libs/tslibs/strptime.pyx [hunk 10]
```diff
@@ -60,15 +62,23 @@ cdef bint format_is_iso(f: str):
...
+    iso_regex = re.compile(
+        r"""
+        ^                     # start of string
+        %Y                    # Year
+        (?:([-/ \\.]?)%m      # month with or without separators
+        (?: \1%d              # day with same separator as for year-month
+        (?:[ T]%H             # hour with separator
+        (?:\:%M               # minute with separator
+        (?:\:%S               # second with separator
+        (?:%z|\.%f(?:%z|Z)?   # timezone or fractional second
```
Comment (MarcoGorelli) [inline pandas/_libs/tslibs/strptime.pyx:74]: `Z` shouldn't be there either
```suggestion
        (?:%z|\.%f(?:%z)?     # timezone or fractional second
```

**Comment by WillAyd** [comment]
lgtm. Nice performance boost. Anything else left to do or want to take out of draft status?

**Comment by MarcoGorelli** [comment]
@ShashwatAgrawal20 if this is ready, could you mark the PR as ready for review please? https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-stage-of-a-pull-request#marking-a-pull-request-as-ready-for-review

let's also add a whatsnew note, something like
```
Performance improvement in :func:`to_datetime` when format is given or can be inferred (:issue:`50465`)
```

**Comment by ShashwatAgrawal20** [comment]
> @ShashwatAgrawal20 if this is ready, could you mark the PR as ready for review please? https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-stage-of-a-pull-request#marking-a-pull-request-as-ready-for-review
> 
> let's also add a whatsnew note, something like
> 
> ```
> Performance improvement in :func:`to_datetime` when format is given or can be inferred (:issue:`50465`)
> ```

`doc/source/whatsnew/v2.0.0.rst` here ?

**Comment by ShashwatAgrawal20** [comment]
> That's right! This is better than the original btw, because this would fail on`'%Y-%m-%d %H:'`, whereas the original would have accepted it
> 
> Leaving open a bit in case there's objections, then will merge

wouldn't have been possible without your guidance.

**Review by MarcoGorelli — APPROVED** [review]
thanks @ShashwatAgrawal20
